### PR TITLE
Twitter authentication

### LINKED ISSUE #29: As an unauthenticated user I can sign up for an account.

### LINKED ISSUE #30: An an registered user, I can sign in with twitter

### PR DESCRIPTION
**Pull request by paulgrever**
closes #29 
closes #30

### CODE REVIEW

File: config/initializers/omniauth.rb [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Rails.application.config.middleware.use OmniAuth::Builder do
+  provider :twitter, ENV['twitter_key'], ENV['twitter_secret']
```
Comment (paulgrever) [inline config/initializers/omniauth.rb]: This should be this way.

**Comment by paulgrever** [comment]
@indiesquidge  @kpearson  ready to  be reviewed by yall!

File: Gemfile [hunk 2]
```diff
@@ -19,6 +19,9 @@ gem 'd3_rails'
 gem "figaro"
 gem 'twilio-ruby', '~> 3.12'
 gem 'faker'
+gem 'omniauth'
+gem 'omniauth-twitter'
+
```
Comment (indiesquidge) [inline Gemfile]: This should be double quotes. We need to keep that consistent.

**Comment by indiesquidge** [comment]
You need to not include binary files from `.DS_Store`. Add it to your global `.gitconfig` or find a way to remove it from OS X completely.

**Comment by indiesquidge** [comment]
You also need to fix those Hound warnings before we should review it, yo.

File: app/controllers/sessions_controller.rb [hunk 3]
```diff
@@ -3,14 +3,24 @@ def new
...
-      session[:user_id] = user.id
-      flash[:success] = "Welcome, #{user.full_name}."
+    if params[:session]
+      user = User.find_by_email(params[:session][:email])
+      if user && user.authenticate(params[:session][:password])
+        session[:user_id] = user.id
+        flash[:success] = "Welcome, #{user.full_name}."
+      else
+        flash[:danger] = "Invalid login."
+      end
```
Comment (indiesquidge) [inline app/controllers/sessions_controller.rb]: These conditionals are hard to follow. We should abstract them into their own private methods.

File: app/controllers/sessions_controller.rb [hunk 4]
```diff
@@ -3,14 +3,24 @@ def new
...
     else
-      flash[:danger] = 'Invalid login.'
+      user = User.find_or_create_from_auth(request.env["omniauth.auth"])
+      if user
+        session[:user_id] = user.id
+        flash[:success] = "Welcome, #{user.full_name}."
+      else
+        flash[:danger] = "Invalid login."
+      end
```
Comment (indiesquidge) [inline app/controllers/sessions_controller.rb]: Same goes for this

File: app/controllers/sessions_controller.rb [hunk 5]
```diff
@@ -3,14 +3,24 @@ def new
...
     end
-    redirect_to(:back)
+    redirect_to items_path
```
Comment (indiesquidge) [inline app/controllers/sessions_controller.rb]: Why are we always taking them back to the items index? Shouldn't we take them back to wherever they logged in from?

File: app/models/user.rb [hunk 6]
```diff
@@ -1,10 +1,23 @@
...
   has_many :order_items, through: :orders
   has_many :items, through: :order_items
   enum role: %w(default admin)
+
+  def self.find_or_create_from_auth(auth)
+    user = User.find_or_create_by(provider: auth.provider, uid: auth.uid)
+    user.provider = auth.provider
+    user.uid = auth.uid
+    user.full_name = auth.info.name
+    user.password = auth.credentials.token
```
Comment (indiesquidge) [inline app/models/user.rb]: I know that we all agreed on this as our solution, but we should ask about it. It feels very hacky to me.

File: config/initializers/omniauth.rb (hunk 1 above)
Comment (indiesquidge) [inline config/initializers/omniauth.rb]: That's fine.

File: spec/factories/user.rb [hunk 7]
```diff
@@ -5,6 +5,8 @@
     display_name "jgu"
     password "gu"
     role "default"
+    provider "twitter"
+    uid "abc123"
```
Comment (indiesquidge) [inline spec/factories/user.rb]: Ehh, these are kind of strange factories. Its not a problem (more of a nitpick), but we should try and keep their fields to something that will be universally understood. Perhaps something like:

```ruby
     email "default.user@gmail.com"
     full_name  "Default User"
     display_name "default_user"
     password "password"
     role "default"
```

Just a thought. It's not a high priority, but it's a quick fix.

File: spec/features/user/signup_spec.rb [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    fill_in("name", with: "")
+    fill_in("email", with: "test@example.com")
+    fill_in("password", with: "test")
+    click_button('Register')
+    expect(User.count).to eq(0)
+    expect(page).to have_content("Input is invalid, please try again")
+  end
+
+  it "can't sign up without a passowrd" do
+    visit root_path
```
Comment (indiesquidge) [inline spec/features/user/signup_spec.rb]: I know that we talked about not having `before` blocks on tests since it is not as self contained, but since visiting the root_path is not essential to how a user signs in, you could move that to a before block. Or you could leave it. Up to you.

File: spec/features/user/signup_spec.rb [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+RSpec.describe "user" do
+  it "can sign up with email" do
+    visit root_path
+    click_link_or_button('Sign up')
+    expect(current_path).to eq(new_user_path)
+    fill_in("name", with: "Paul")
+    fill_in("email", with: "test@example.com")
+    fill_in("password", with: "test")
+    click_button('Register')
+    expect(User.count).to eq(1)
```
Comment (indiesquidge) [inline spec/features/user/signup_spec.rb]: I would also test where it takes the user after they click register as well. Just a simple `expect` line.

File: spec/features/user/signup_spec.rb [hunk 10]
```diff
@@ -0,0 +1,50 @@
+require 'rails_helper'
+
+RSpec.describe "user" do
+  it "can sign up with email" do
+    visit root_path
+    click_link_or_button('Sign up')
+    expect(current_path).to eq(new_user_path)
```
Comment (indiesquidge) [inline spec/features/user/signup_spec.rb]: This isn't that necessary. You can leave it, but it's probably implied since the user clicked a button to sign up. It would fail on the `fill_in` part if we weren't on that path.

File: spec/features/user/signup_spec.rb [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    expect(page).to have_content("Input is invalid, please try again")
+  end
+
+  it "can't sign up without a passowrd" do
+    visit root_path
+    click_link_or_button('Sign up')
+    expect(current_path).to eq(new_user_path)
+    fill_in("name", with: "Paul")
+    fill_in("email", with: "test@example.com")
+    fill_in("password", with: "")
```
Comment (indiesquidge) [inline spec/features/user/signup_spec.rb]: Based on these tests, we are only requiring new users to sign up with a name, email, and password, yeah? That needs to be implied somewhere on the user view (like the little `*` you sometimes see on forms when the field is required). Otherwise we have a giant signup form and the user won't know what is required and what is not.

File: app/controllers/sessions_controller.rb (hunk 5 above)
Comment (paulgrever) [inline app/controllers/sessions_controller.rb]: @indiesquidge  because when users login with twitter it takes them to a twitter page, so it doesn't have to be items, but we can't send them back to the twitter page as it's a different site.

File: app/models/user.rb (hunk 6 above)
Comment (paulgrever) [inline app/models/user.rb]: Yes, we can refactor at a later time.

File: spec/factories/user.rb (hunk 7 above)
Comment (paulgrever) [inline spec/factories/user.rb]: I agree with your point but I don't think it's a good use of resources and time as it will break tests I believe.

File: spec/features/user/signup_spec.rb (hunk 11 above)
Comment (paulgrever) [inline spec/features/user/signup_spec.rb]: I didn't do much styling as I thought our long term goal was to make this a modal? 

If we aren't going to go that way, I will fix these.

File: app/models/user.rb [hunk 12]
```diff
@@ -1,10 +1,23 @@
 class User < ActiveRecord::Base
   validates :full_name, presence: true
-  validates :email, format: { with: /\A([^@\s]+)@((?:[-a-z0-9]+\.)+[a-z]{2,})\Z/i }
-  validates :display_name, length: { maximum: 32, minimum: 2 }, presence: false
+  validates :email,
+             format: { with: /\A([^@\s]+)@((?:[-a-z0-9]+\.)+[a-z]{2,})\Z/i }
```
Comment (paulgrever) [inline app/models/user.rb]: Since this a long regex, if you break the fields 63 test fail. Not sure how to fix that, via a new line.

File: config/initializers/omniauth.rb [hunk 13]
```diff
@@ -0,0 +1,3 @@
...
+end
```
Comment (paulgrever) [inline config/initializers/omniauth.rb]: I added a new line... it's weird that it's not registering.

Comment (paulgrever) [inline config/initializers/omniauth.rb]: But it's actually fixed in the code... Super strange.

File: spec/factories/user.rb (hunk 7 above)
Comment (indiesquidge) [inline spec/factories/user.rb]: Haha it would take two seconds but it's fine

**Comment by kpearson** [comment]
Nice work!
